File: flink-scala-code/src/main/scala/com/atguigu/day2/MultiStreamTransformations.scala
package com.atguigu.day2

import org.apache.flink.streaming.api.functions.co.CoFlatMapFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

// freeBSD: Bill Joy: bsd, vi, tcpip, sun公司 -> MacOS
// Linux
object MultiStreamTransformations {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val keyedTempStream = env.addSource(new SensorSource).keyBy(r => r.id)

    val smokeLevelStream = env.fromElements("LOW", "HIGH").setParallelism(1)

    keyedTempStream
      .connect(smokeLevelStream.broadcast)
      .flatMap(new RaiseAlertFlatMap)
      .print()

    env.execute()
  }

  class RaiseAlertFlatMap extends CoFlatMapFunction[SensorReading, String, Alert] {
    private var smokeLevel: String = "LOW"

    override def flatMap1(value: SensorReading, out: Collector[Alert]): Unit = {
      if (smokeLevel == "HIGH" && value.temperature > -10000.0) {
        out.collect(Alert(value.toString, value.timestamp))
      }
    }

    override def flatMap2(value: String, out: Collector[Alert]): Unit = {
      smokeLevel = value
    }
  }
}

File: flink-code-java-0421/src/main/java/com/atguigu/day02/MultiStreamTransformations.java
package com.atguigu.day02;

import com.atguigu.day02.util.*;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.co.CoFlatMapFunction;
import org.apache.flink.util.Collector;

public class MultiStreamTransformations {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStream<SensorReading> tempReadings = env.addSource(new SensorSource());

        // 并行度设置为1
        DataStream<SmokeLevel> smokeReadings = env.addSource(new SmokeLevelSource()).setParallelism(1);

        tempReadings.keyBy(r -> r.id)
                .connect(smokeReadings.broadcast())
                .flatMap(new RaiseAlertFlatMap())
                .print();

        env.execute();
    }

    public static class RaiseAlertFlatMap implements CoFlatMapFunction<SensorReading, SmokeLevel, Alert> {
        private SmokeLevel smokeLevel = SmokeLevel.LOW;

        @Override
        public void flatMap1(SensorReading sensorReading, Collector<Alert> collector) throws Exception {
            if (this.smokeLevel == SmokeLevel.HIGH && sensorReading.temperature > 0.0) {
                collector.collect(new Alert("报警！" + sensorReading, sensorReading.timestamp));
            }
        }

        @Override
        public void flatMap2(SmokeLevel smokeLevel, Collector<Alert> collector) throws Exception {
            this.smokeLevel = smokeLevel;
        }
    }
}
